Creates the `action:create` command

weierophinney opened the pull request:
Essentially, it's the same command as `handler:create`, just with a different name during instantiation. When `action:create` is used, the required argument is renamed to `action` as well.

Fixes #51

weierophinney committed:
Creates the `action:create` command

Essentially, it's the same command as `handler:create`, just with a
different name during instantiation. When `action:create` is used, the
required argument is renamed to `action` as well.

michalbundyra commented on src/CreateHandler/CreateHandlerCommand.php:
```diff
@@ -59,9 +59,13 @@ class. For a path, it matches the class namespace against PSR-4 autoloader
      */
     protected function configure() : void
     {
+        $this->handlerArgument = (bool) preg_match('/^action:/', $this->getName())
```
We don't need to use regexp, `strpos === 0` is enough.

michalbundyra replied:
I can't see property `handlerArgument`, should be added instead of dynamic creation.

weierophinney replied:
It's needed withing the `execute()` method, so that we pull the correct argument.

michalbundyra replied:
Yes, but what I ment is that property `handlerArgument` is not defined in the class (it is created here dynamically what is not a good practice) and I've suggested to use `strpos` instead of regular expression, because there is no reason to have the regexp here. See my comment below also how I would change it with property:
```php
protected $handlerArgument = 'handler';
```
(or `private`).

michalbundyra commented on src/CreateHandler/CreateHandlerCommand.php:
```diff
@@ -59,9 +59,13 @@ class. For a path, it matches the class namespace against PSR-4 autoloader
     {
+        $this->handlerArgument = (bool) preg_match('/^action:/', $this->getName())
+            ? 'action'
+            : 'handler'; // default argument
```
If we add the property maybe there we can have the default value and change it here only:
```php
if (strpos($this->getName(), 'action:') === 0) {
    $this->handlerArgument = 'action';
}
```

michalbundyra commented on src/CreateHandler/CreateHandlerCommand.php:
```diff
@@ -73,7 +77,7 @@ protected function configure() : void
-        $handler = $input->getArgument('handler');
+        $handler = $input->getArgument($this->handlerArgument);
 
         $output->writeln(sprintf('<info>Creating request handler %s...</info>', $handler));
```
Hm, maybe we should then also update message? And help? Hm... (?)

weierophinney replied:
It's still creating a _request handler_; the only difference is in the command name and the argument name...

Which is why I'm not particularly convinced this patch is really needed.

@akrabat — what are your thoughts? You can still call them actions even using the `handler:create` command:

```bash
$ ./bin/expressive handler:create "App\Action\HelloWorldAction"
```

The question is really: do we also really want/need an `action:create` command, and, if so, how far should we go towards using "action" terminology in that command? Should the term "request handler" be omitted entirely from help descriptions?

akrabat replied:
It's all down to how accessible this is supposed to be to people who don't know PSR-15 and how "compatible" you want the terminology to be with other frameworks.

Personally, I think that they are actions that happen to implement PSR-15's `RequestHandlerInterface`.

If you want all users to think "Request handler", then this PR isn't needed.

akrabat replied:
Oh and personally,  I'd change message to say "Creating action"

I would have kept the namespace as Action too, so maybe I'm not on the same page though.

michalbundyra commented on test/CreateHandler/CreateHandlerCommandTest.php:
```diff
@@ -100,6 +100,16 @@ public function testConfigureSetsExpectedArguments()
+    {
+        $command = new CreateHandlerCommand('action:create');
+        $definition = $command->getDefinition();
+        $this->assertTrue($definition->hasArgument('action'));
```
Maybe also $this->assertFalse($definition->hasArgument('handler'));` ?

weierophinney committed:
Provides different help messages when requesting an action

This patch modifies the `CreateHandlerCommand` to configure the command
differently based on whether an action is requested; in such a case, it
uses a completely different set of help messages and descriptions that
reference "actions" and "action classes" instead of "request handlers".

It also adds a template for the initial status message.
